### PR TITLE
Add lock for MutableTree.unsavedAdditons and MutableTree.unsavedRemovals

### PR DESCRIPTION
**Pull request by giskook**
add unittest for the concurrent set and get

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: libs/iavl/nodedb_update_branch.go [hunk 1]
```diff
@@ -129,14 +129,9 @@ func updateBranchAndSaveNodeToChan(node *Node, saveNodesCh chan<- *Node) []byte
...
+func (ndb *nodeDB) updateBranchForFastNode(fnc *fastNodeChanges) {
 	ndb.mtx.Lock()
-	for k, v := range additions {
-		ndb.prePersistFastNode.add(k, v)
-	}
-	for k, v := range removals {
-		ndb.prePersistFastNode.remove(k, v)
-	}
+	ndb.prePersistFastNode.mergeLater(fnc)
 	ndb.mtx.Unlock()
```
Comment (lyh169) [inline libs/iavl/nodedb_update_branch.go:135]: this location should  use `mergeLater()` inside Lock ?

File: libs/iavl/nodedb_fnc.go [hunk 2]
```diff
@@ -64,40 +90,72 @@ func (fnc *fastNodeChanges) clone() *fastNodeChanges {
...
+		if !fnc.checkAdditions(amino.StrToBytes(k)) && !fnc.checkRemovals(amino.StrToBytes(k)) {
+			fnc.remove(amino.StrToBytes(k), v)
 		}
 	}
 	return fnc
 }
 
+// mergeLater merge later to fnc, later is new than fnc
+func (fnc *fastNodeChanges) mergeLater(later *fastNodeChanges) {
+	for k, v := range later.additions {
```
Comment (lyh169) [inline libs/iavl/nodedb_fnc.go:140]: this merge shoud not judge the fnc the both add and removal ?

Comment (giskook) [reply libs/iavl/nodedb_fnc.go:140]: the `later` fastNodeChanges is newest than `fnc` fastNodeChanges. we should consider `later` first. so we do not check additions and removals

File: libs/iavl/nodedb_update_branch.go (hunk 1 above)
Comment (giskook) [reply libs/iavl/nodedb_update_branch.go:135]: this `ndb.mtx` works for `ndb.prePersistFastNode`. besides `mergeLater` dont need the `fastNodeChanges`' lock. as when we commit the execution is sequential. so here `read` the map will not conflict with `write`.

**Review by lyh169 — DISMISSED** [review]

**Review by cwbhhjl — DISMISSED** [review]

**Review by xiangjianmeng — APPROVED** [review]

**Pull request merged**